### PR TITLE
My Site Dashboard: Avoid showing recently deleted posts in dashboard

### PR DESCRIPTION
**Pull request by hassaanelgarem**
Part of #18308

## Description
This PR fixes an issue where it was possible to edit a post after trashing it.
It was mentioned in https://github.com/wordpress-mobile/WordPress-iOS/pull/18351#issuecomment-1097115823

## Testing Instructions
1. Open a site with draft posts
2. Make sure you're on a slow connection
3. Delete a draft post
4. Go back immediately to the dashboard
5. Make sure the post is not displayed.

## Regression Notes
1. Potential unintended areas of impact
    - Posts dashboard cards
    - Posts list view controller

3. What I did to test those areas of impact (or what existing automated tests I relied on)
    - Tested manually and added relevant unit tests

4. What automated tests I added (or what prevented me from doing so)
    - `testDraftFilterDoesNotincludeRecentlyTrashedPosts`
    - `testTrashingAPostWillUpdateItsRevisionStatusAfterSyncProperty`

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by momo-ozawa — COMMENTED** [review]
@hassaanelgarem 👋 
I found a few issues while testing.

### Core Data crash 

This isn't 100% reproducible, but I've seen this crash twice while testing.

```
2022-04-18 11:25:59:830 WordPress[49166:1128620] Generating exception with reason:
Non-optional property with a nil value on Category.blog
2022-04-18 11:25:59.851340+0100 WordPress[49166:1128620] *** Terminating app due to uncaught exception 'Unresolved Core Data save error (Main Context)', reason: 'Non-optional property with a nil value on Category.blog'
```
- Turn on the "Very bad network" profile for Network Link Conditioner
- Switch to a blog with multiple draft posts
- Trash all the draft posts
- 💥 Crash due to Core Data error


### Create your first post / next post doesn't show immediately after trashing post

- Turn on the "Very bad network" profile for Network Link Conditioner
- Switch to a site that displays the "Create your first post" or "Create your next post card"
- Create a draft post
- Trash the draft post you just created
- Go back to the dashboard
- 🐛 It takes a while to display the correct card OR it gets stuck on an empty card state (where only the "Create a draft post" header is showing)


https://user-images.githubusercontent.com/6711616/163797035-468d180a-3295-46f9-b479-a9460a4e9be0.mp4

File: WordPress/Classes/Services/PostService.m [hunk 1]
```diff
@@ -604,6 +604,8 @@ - (void)trashRemotePostWithPost:(AbstractPost*)post
     
     RemotePost *remotePost = [self remotePostWithPost:post];
     id<PostServiceRemote> remote = [self.postServiceRemoteFactory forBlog:post.blog];
+    post.statusAfterSync = PostStatusTrash;
+    post.latest.statusAfterSync = PostStatusTrash;
```
Comment (leandroalonso) [inline WordPress/Classes/Services/PostService.m:608]: I'm a little bit concerned about this change because we're changing `statusAfterSync` before the sync (which is the `trashPost` below).

Ideally, we would have another property on the post to identify that the post has been trashed locally — but I think pursuing that involves Core Data and some other changes that are out of scope for this task.

Plus, on `failureBlock` below we would have to handle resetting `statusAfterSync` in case it fails.

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewModel.swift [hunk 2]
```diff
@@ -30,7 +30,7 @@ class PostsCardViewModel: NSObject {
 
     private let postService: PostService
 
-    private var postListFilter: PostListFilter = PostListFilter.draftFilter()
+    private var postListFilter: PostListFilter = PostListFilter.draftFilter(includeRecentlyTrashed: false)
```
Comment (leandroalonso) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewModel.swift:33]: I guess it probably would be good to keep the posts list and postcard in sync in terms of how they work, this is probably going to make things simple.

**Review by leandroalonso — COMMENTED** [review]
@hassaanelgarem it all works as described!

However, I have a few concerns about moving on with this change, which I explained in the comments.

tl;dr: posts are complex.

When I mentioned this issue I didn't realize it's an issue that it's already on the app — it wasn't added with the dashboard.

If you delete a post on a bad connection, leave the posts list and go back, the post will be there too, which is a similar issue.

We can probably follow a path like the one you started — optimistically trashing the post but we would also need to deal with failures. When it fails, what do we do? Just reset the status to what it was before? Do nothing? (but then the post hasn't been deleted)

I think this can possibly open a can of worms so it might be better to leave as it is and wait for feedback (if there's any). But up to you! Let me know what you think. :)

**Comment by hassaanelgarem** [comment]
@leandroalonso These are 100% valid concerns. I was a bit wary of making these changes as well.

Given all this, I agree with you that it's best to leave this for now. I'll close this PR and create an issue with the existing issues (dashboard and posts list). And maybe we can try tackling this again in a better way in the near future.

**Comment by hassaanelgarem** [comment]
@momo-ozawa Thanks for the review! Unfortunately, I'll be closing this PR for now. 
Sorry if I wasted your time 🙈
